README.md
# initscripts
This repository contains source code for **legacy** *System V [initscripts](https://en.wikipedia.org/wiki/Init)*,
which are primarily used in *[Linux](https://en.wikipedia.org/wiki/Linux) distributions like e.g.*:
* [Fedora](https://en.wikipedia.org/wiki/Fedora_(operating_system))
* [Red Hat Enterprise Linux](https://en.wikipedia.org/wiki/Red_Hat_Enterprise_Linux)
* [CentOS](https://en.wikipedia.org/wiki/CentOS)
* and some others as well...


ChangeLog
commit f8c758709c2dea6651afcc75fa7cceb3b7ff7afc
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 18:17:22 2010 +0200

    version 8.45.32

 initscripts.spec |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 967b64ce3f9754f7de902ed290c813634c182803
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 18:16:21 2010 +0200

    52a7a1f115bb87323620b73f5dde21e2c5e7d9a7

 ChangeLog                        |    7 ++++---
 udev/rules.d/54-cu3088-fix.rules |   10 ++++++++++
 2 files changed, 14 insertions(+), 3 deletions(-)

commit c014f9c6c09b50302e89f9df21077343b6c363b8
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 17:34:33 2010 +0200

    initrscripts.spec: sync with cvs repo

 initscripts.spec |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 923ef28cc4d4654e58033c81cd44ce9e1e4734b8
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 17:30:23 2010 +0200

    version 8.45.31

 ChangeLog        | 1016 ++++++++++++++++++++++++++++++++++--------------------
 initscripts.spec |    4 +-
 2 files changed, 636 insertions(+), 384 deletions(-)

commit 52a7a1f115bb87323620b73f5dde21e2c5e7d9a7
Author: Phil Knirsch <pknirsch@redhat.com>
Date:   Mon Sep 13 16:29:45 2010 +0200

    s390: Added required 54-cu3088-fix.rules udev rules file (#475721)

 initscripts.spec |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

commit b74c943aa01f6a8facc36246dc20ccebd9eb690c
Author: David Cantrell <dcantrell@redhat.com>
Date:   Fri Sep 26 16:41:21 2008 -1000

    Expand get_hostname() function to support IPv4 and IPv6.
    
    Make sure that get_hostname() if the user specifies either address
    family.

 src/ipcalc.c |   36 +++++++++++++++++++++++++++++-------
 1 files changed, 29 insertions(+), 7 deletions(-)

commit f085cdfdc0d67fe240c23e75d14dce6534d36c8a
Author: David Cantrell <dcantrell@redhat.com>
Date:   Fri Sep 26 16:30:38 2008 -1000

    Remove unnecessary tab.

 src/ipcalc.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit e921ac88353159dee451f7c0094fca95a9b2bb56
Author: David Cantrell <dcantrell@redhat.com>
Date:   Fri Sep 26 16:29:21 2008 -1000

    Correct address family option handling in ipcalc.
    
    Make sure the default is IPv4.  If the check option is specified,
    make sure we just validate the address and return 1 or 0.

 src/ipcalc.c |   83 ++++++++++++++++++++++++++++-----------------------------
 1 files changed, 41 insertions(+), 42 deletions(-)

commit 9253ed4bdddd635706a0a741259cbaa8a21d8b89
Author: David Cantrell <dcantrell@redhat.com>
Date:   Fri Sep 26 16:26:27 2008 -1000

    Compile ipcalc with _GNU_SOURCE defined (for asprintf).

 src/Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 192392a47594d8141aaf14a0a705bc421899adb0
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 16:20:36 2010 +0200

    remove deprecation comments for raw devices and fix path

 initscripts.spec       |    6 ++++++
 rc.d/init.d/rawdevices |    6 ++----
 sysconfig/rawdevices   |    3 ---
 3 files changed, 8 insertions(+), 7 deletions(-)

commit 6c0c324a6c0c031ed4604c6a020829503e8d086f
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 15:59:40 2010 +0200

    version 8.45.30

 initscripts.spec |   28 +++++++++++++++++++++++++++-
 1 files changed, 27 insertions(+), 1 deletions(-)

commit 8f65b61fc53d245c7865bfa0ab879337d9c6ff31
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 15:48:34 2010 +0200

    rc.sysinit: also remove /var/run/dovecot/*
    
    https://bugzilla.redhat.com/show_bug.cgi?id=492994

 rc.d/rc.sysinit |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 85f809ad1d6f47d4fe73a07fa358fd917bc94b87
Author: Bill Nottingham <notting@redhat.com>
Date:   Wed Sep 2 16:28:02 2009 -0400

    Doucment DHCP_HOSTNAME (#498052)

 sysconfig.txt |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

commit 52dfb70cd860e4d0fd51e37e30a081fd197f1b74
Author: Bill Nottingham <notting@redhat.com>
Date:   Fri Jul 20 20:54:47 2007 +0000

    move quotacheck to after filesystem mounting so it actually works (#249003, <tometzky@batory.org.pl>)
    also, nuke the convertquota stuff that has been obsolete (and not working) for quite some time

 rc.d/rc.sysinit |   25 +++++--------------------
 1 files changed, 5 insertions(+), 20 deletions(-)

commit 873660981a5a42a6de411e04958a4f204704efe1
Author: Bill Nottingham <notting@redhat.com>
Date:   Tue Dec 15 10:59:44 2009 -0500

    Silence error. (#516569)

 sysconfig/network-scripts/network-functions |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 3ae2380e43092dfdc12b2d242ade6bd1379fe1e7
Author: Bill Nottingham <notting@redhat.com>
Date:   Tue Oct 20 16:56:32 2009 -0400

    Load bonding driver if BONDING_OPTS is defined. (#516569)

 sysconfig/network-scripts/network-functions |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

commit f08000519addd0b13fada0eba4afef72871613c1
Author: Bill Nottingham <notting@redhat.com>
Date:   Thu Aug 27 13:25:41 2009 -0400

    Restore context of recreated files. (#519748)

 rc.d/rc.sysinit |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

commit 209e5e18cbd1cbb76f7e49c83854763c4e9f1d1c
Author: Bill Nottingham <notting@redhat.com>
Date:   Fri Oct 9 16:18:04 2009 -0400

    Do not try to activate subsets and layer raid and handle 'no raid disks'.
    
    - 'no raid disks' handling: (#527726, <mschmidt@redhat.com>)
    - no subsets: https://bugzilla.redhat.com/show_bug.cgi?id=524356

 rc.d/rc.sysinit |   15 +++++++--------
 1 files changed, 7 insertions(+), 8 deletions(-)

commit 40ddde4c4e6ad1dc855a8a3f4ae594d6c2d16179
Author: Bill Nottingham <notting@redhat.com>
Date:   Wed Oct 7 11:10:12 2009 -0400

    Add /var/lib/xend. (#526046)

 rwtab |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

commit 643c05c59a581663140e307f08732c6bef3d7622
Author: Bill Nottingham <notting@redhat.com>

src/ipcalc.c
/*
 * Copyright (c) 1997-2009 Red Hat, Inc. All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 *
 * Authors:
 *   Erik Troan <ewt@redhat.com>
 *   Preston Brown <pbrown@redhat.com>
 *   David Cantrell <dcantrell@redhat.com>
 */

#include <ctype.h>
#include <popt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>

/*!
  \file ipcalc.c
  \brief provides utilities for manipulating IP addresses.

  ipcalc provides utilities and a front-end command line interface for
  manipulating IP addresses, and calculating various aspects of an ip
  address/netmask/network address/prefix/etc.

  Functionality can be accessed from other languages from the library
  interface, documented here.  To use ipcalc from the shell, read the
  ipcalc(1) manual page.

  When passing parameters to the various functions, take note of whether they
  take host byte order or network byte order.  Most take host byte order, and
  return host byte order, but there are some exceptions.
*/

/*!
  \fn struct in_addr prefix2mask(int bits)
  \brief creates a netmask from a specified number of bits

  This function converts a prefix length to a netmask.  As CIDR (classless
  internet domain internet domain routing) has taken off, more an more IP
  addresses are being specified in the format address/prefix
  (i.e. 192.168.2.3/24, with a corresponding netmask 255.255.255.0).  If you
  need to see what netmask corresponds to the prefix part of the address, this
  is the function.  See also \ref mask2prefix.

  \param prefix is the number of bits to create a mask for.
  \return a network mask, in network byte order.
*/
struct in_addr prefix2mask(int prefix) {
    struct in_addr mask;
    memset(&mask, 0, sizeof(mask));
    if (prefix) {
        mask.s_addr = htonl(~((1 << (32 - prefix)) - 1));
    } else {
        mask.s_addr = htonl(0);
    }
    return mask;
}

/*!
  \fn int mask2prefix(struct in_addr mask)
  \brief calculates the number of bits masked off by a netmask.

  This function calculates the significant bits in an IP address as specified by
  a netmask.  See also \ref prefix2mask.

  \param mask is the netmask, specified as an struct in_addr in network byte order.
  \return the number of significant bits.  */
int mask2prefix(struct in_addr mask)
{
    int count;
    uint32_t saddr = ntohl(mask.s_addr);

    for (count=0; saddr > 0; count++) {
        saddr=saddr << 1;
    }

    return count;
}

/*!
  \fn struct in_addr default_netmask(struct in_addr addr)

  \brief returns the default (canonical) netmask associated with specified IP
  address.

  When the Internet was originally set up, various ranges of IP addresses were
  segmented into three network classes: A, B, and C.  This function will return
  a netmask that is associated with the IP address specified defining where it
  falls in the predefined classes.

  \param addr an IP address in network byte order.
  \return a netmask in network byte order.  */
struct in_addr default_netmask(struct in_addr addr)
{
    uint32_t saddr = addr.s_addr;
    struct in_addr mask;

    memset(&mask, 0, sizeof(mask));

    if (((ntohl(saddr) & 0xFF000000) >> 24) <= 127)
        mask.s_addr = htonl(0xFF000000);
    else if (((ntohl(saddr) & 0xFF000000) >> 24) <= 191)
        mask.s_addr = htonl(0xFFFF0000);
    else
        mask.s_addr = htonl(0xFFFFFF00);

    return mask;
}

/*!
  \fn struct in_addr calc_broadcast(struct in_addr addr, int prefix)

  \brief calculate broadcast address given an IP address and a prefix length.

  \param addr an IP address in network byte order.
  \param prefix a prefix length.

  \return the calculated broadcast address for the network, in network byte
  order.
*/
struct in_addr calc_broadcast(struct in_addr addr, int prefix)
{
    struct in_addr mask = prefix2mask(prefix);
    struct in_addr broadcast;

    memset(&broadcast, 0, sizeof(broadcast));
    broadcast.s_addr = (addr.s_addr & mask.s_addr) | ~mask.s_addr;
    return broadcast;
}

/*!
  \fn struct in_addr calc_network(struct in_addr addr, int prefix)
  \brief calculates the network address for a specified address and prefix.

  \param addr an IP address, in network byte order
  \param prefix the network prefix
  \return the base address of the network that addr is associated with, in
  network byte order.
*/
struct in_addr calc_network(struct in_addr addr, int prefix)
{
    struct in_addr mask = prefix2mask(prefix);
    struct in_addr network;

    memset(&network, 0, sizeof(network));
    network.s_addr = addr.s_addr & mask.s_addr;
    return network;
}

/*!
  \fn const char *get_hostname(int family, void *addr)
  \brief returns the hostname associated with the specified IP address

  \param family the address family, either AF_INET or AF_INET6.
  \param addr an IP address to find a hostname for, in network byte order,
  should either be a pointer to a struct in_addr or a struct in6_addr.

  \return a hostname, or NULL if one cannot be determined.  Hostname is stored
  in a static buffer that may disappear at any time, the caller should copy the
  data if it needs permanent storage.
*/
char *get_hostname(int family, void *addr)
{
    struct hostent * hostinfo = NULL;
    int x;
    struct in_addr addr4;
    struct in6_addr addr6;

    if (family == AF_INET) {
        memset(&addr4, 0, sizeof(addr4));
        memcpy(&addr4, addr, sizeof(addr4));
        hostinfo = gethostbyaddr((const void *) &addr4,
                                 sizeof(addr4), family);
    } else if (family == AF_INET6) {
        memset(&addr6, 0, sizeof(addr6));
        memcpy(&addr6, addr, sizeof(addr6));
        hostinfo = gethostbyaddr((const void *) &addr6,
                                 sizeof(addr6), family);
    }

    if (!hostinfo)
        return NULL;

    for (x=0; hostinfo->h_name[x]; x++) {
        hostinfo->h_name[x] = tolower(hostinfo->h_name[x]);
    }
    return hostinfo->h_name;
}

/*!
  \fn main(int argc, const char **argv)
  \brief wrapper program for ipcalc functions.

  This is a wrapper program for the functions that the ipcalc library provides.
  It can be used from shell scripts or directly from the command line.

  For more information, please see the ipcalc(1) man page.
*/
int main(int argc, const char **argv) {
    int showBroadcast = 0, showPrefix = 0, showNetwork = 0;
    int showHostname = 0, showNetmask = 0;
    int beSilent = 0;
    int doCheck = 0, familyIPv4 = 0, familyIPv6 = 0;
    int rc;
    poptContext optCon;
    char *ipStr, *prefixStr, *netmaskStr, *chptr;
    char *hostName = NULL;
    char namebuf[INET6_ADDRSTRLEN+1];
    struct in_addr ip, netmask, network, broadcast;
    struct in6_addr ip6;
    int prefix = -1;
    char errBuf[250];
    struct poptOption optionsTable[] = {
        { "check", 'c', 0, &doCheck, 0,
          "Validate IP address for specified address family", },
        { "ipv4", '4', 0, &familyIPv4, 0,
          "IPv4 address family (default)", },
        { "ipv6", '6', 0, &familyIPv6, 0,
          "IPv6 address family", },
        { "broadcast", 'b', 0, &showBroadcast, 0,
          "Display calculated broadcast address", },
        { "hostname", 'h', 0, &showHostname, 0,
          "Show hostname determined via DNS" },
        { "netmask", 'm', 0, &showNetmask, 0,
          "Display default netmask for IP (class A, B, or C)" },
        { "network", 'n', 0, &showNetwork, 0,
          "Display network address", },
        { "prefix", 'p', 0, &showPrefix, 0,
          "Display network prefix", },
        { "silent", 's', 0, &beSilent, 0,
          "Don't ever display error messages" },
        POPT_AUTOHELP
        { NULL, '\0', 0, 0, 0, NULL, NULL }
    };

    optCon = poptGetContext("ipcalc", argc, argv, optionsTable, 0);
    poptReadDefaultConfig(optCon, 1);

    if ((rc = poptGetNextOpt(optCon)) < -1) {
        if (!beSilent) {
            fprintf(stderr, "ipcalc: bad argument %s: %s\n",
                    poptBadOption(optCon, POPT_BADOPTION_NOALIAS),
                    poptStrerror(rc));
            poptPrintHelp(optCon, stderr, 0);
        }
        return 1;
    }

    if (!(ipStr = (char *) poptGetArg(optCon))) {
        if (!beSilent) {
            fprintf(stderr, "ipcalc: ip address expected\n");
            poptPrintHelp(optCon, stderr, 0);
        }
        return 1;
    }

    /* if there is a : in the address, it is an IPv6 address */
    if (strchr(ipStr,':') != NULL) {
        familyIPv6=1;
    }

    if (strchr(ipStr,'/') != NULL) {
        prefixStr = strchr(ipStr, '/') + 1;
        prefixStr--;
        *prefixStr = '\0';  /* fix up ipStr */
        prefixStr++;
    } else {
        prefixStr = NULL;
    }

    if (prefixStr != NULL) {
        prefix = atoi(prefixStr);
        if (prefix < 0 || ((familyIPv6 && prefix > 128) || (!familyIPv6 && prefix > 32))) {
            if (!beSilent)
                fprintf(stderr, "ipcalc: bad prefix: %s\n", prefixStr);
            return 1;
        }
    }

    if (showBroadcast || showNetwork || showPrefix) {
        if (!(netmaskStr = (char *) poptGetArg(optCon)) && (prefix < 0)) {
            if (!beSilent) {
                fprintf(stderr, "ipcalc: netmask or prefix expected\n");
                poptPrintHelp(optCon, stderr, 0);
            }
            return 1;
        } else if (netmaskStr && prefix >= 0) {
            if (!beSilent) {
                fprintf(stderr, "ipcalc: both netmask and prefix specified\n");
                poptPrintHelp(optCon, stderr, 0);
            }
            return 1;
        } else if (netmaskStr) {
            if (inet_pton(AF_INET, netmaskStr, &netmask) <= 0) {
                if (!beSilent)
                    fprintf(stderr, "ipcalc: bad netmask: %s\n", netmaskStr);
                return 1;
            }
            prefix = mask2prefix(netmask);
        }
    }

    if ((chptr = (char *) poptGetArg(optCon))) {
        if (!beSilent) {
            fprintf(stderr, "ipcalc: unexpected argument: %s\n", chptr);
            poptPrintHelp(optCon, stderr, 0);
        }
        return 1;
    }

    if (!familyIPv4 && !familyIPv6)
        familyIPv4 = 1;

    if (familyIPv4 && familyIPv6) {
        if (!beSilent) {
            fprintf(stderr, "ipcalc: cannot specify both address families\n");
        }
        return 1;
    }

    /* Handle CIDR entries such as 172/8 */
    if (prefix >= 0 && familyIPv4) {
        char *tmp = ipStr;
        int i;

        for (i=3; i> 0; i--) {
            tmp = strchr(tmp,'.');
            if (!tmp)
                break;
            else
                tmp++;
        }

        tmp = NULL;
        for (; i>0; i--) {
            if (asprintf(&tmp, "%s.0", ipStr) == -1) {
                fprintf(stderr, "Memory allocation failure line %d\n", __LINE__);
                abort();
            }
            ipStr = tmp;
        }
    }

    if (familyIPv4) {
        if (inet_pton(AF_INET, ipStr, &ip) <= 0) {
            if (!beSilent)
                fprintf(stderr, "ipcalc: bad IPv4 address: %s\n", ipStr);
            return 1;
        } else if (prefix > 32) {
            if (!beSilent)
                fprintf(stderr, "ipcalc: bad IPv4 prefix %d\n", prefix);
            return 1;
        } else {
            if (doCheck)
                return 0;
        }
    }

    if (familyIPv6) {
        if (inet_pton(AF_INET6, ipStr, &ip6) <= 0) {
            if (!beSilent)
                fprintf(stderr, "ipcalc: bad IPv6 address: %s\n", ipStr);
            return 1;
        } else if (prefix > 128) {
            if (!beSilent)
                fprintf(stderr, "ipcalc: bad IPv6 prefix %d\n", prefix);
            return 1;
        } else {
            if (doCheck)
                return 0;
        }
    }

    if (familyIPv6 &&
        (showBroadcast || showNetmask || showNetwork || showPrefix)) {
        if (!beSilent) {
            fprintf(stderr, "ipcalc: unable to show setting for IPv6\n");
        }
        return 1;
    }

    if (familyIPv4 &&
        !(showNetmask|showPrefix|showBroadcast|showNetwork|showHostname)) {
        poptPrintHelp(optCon, stderr, 0);
        return 1;
    }

    poptFreeContext(optCon);

    /* we know what we want to display now, so display it. */

    if (showNetmask) {
        if (prefix >= 0) {
            netmask = prefix2mask(prefix);
        } else {
            netmask = default_netmask(ip);
            prefix = mask2prefix(netmask);
        }

        memset(&namebuf, '\0', sizeof(namebuf));

        if (inet_ntop(AF_INET, &netmask, namebuf, INET_ADDRSTRLEN) == NULL) {
            fprintf(stderr, "Memory allocation failure line %d\n", __LINE__);
            abort();
        }

        printf("NETMASK=%s\n", namebuf);
    }

    if (showPrefix) {
        if (prefix == -1)
            prefix = mask2prefix(ip);
        printf("PREFIX=%d\n", prefix);
    }

    if (showBroadcast) {
        broadcast = calc_broadcast(ip, prefix);
        memset(&namebuf, '\0', sizeof(namebuf));

        if (inet_ntop(AF_INET, &broadcast, namebuf, INET_ADDRSTRLEN) == NULL) {
            fprintf(stderr, "Memory allocation failure line %d\n", __LINE__);
            abort();
        }

        printf("BROADCAST=%s\n", namebuf);
    }

    if (showNetwork) {
        network = calc_network(ip, prefix);
        memset(&namebuf, '\0', sizeof(namebuf));

        if (inet_ntop(AF_INET, &network, namebuf, INET_ADDRSTRLEN) == NULL) {
            fprintf(stderr, "Memory allocation failure line %d\n", __LINE__);
            abort();
        }

        printf("NETWORK=%s\n", namebuf);
    }

    if (showHostname) {
        if (familyIPv4) {
            hostName = get_hostname(AF_INET, &ip);
        } else if (familyIPv6) {
            hostName = get_hostname(AF_INET6, &ip6);
        }

        if (hostName == NULL) {
            if (!beSilent) {
                sprintf(errBuf, "ipcalc: cannot find hostname for %s", ipStr);
                herror(errBuf);
            }
            return 1;
        }

        printf("HOSTNAME=%s\n", hostName);
    }

    return 0;
}
